BUG/DEPR: loc.__setitem__ incorrectly accepting positional slices

jbrockmendel opened the pull request:
- [x] closes #26412
- [x] closes #16121
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

I plan to flesh out the tests before this goes in.

cc @toobaz

jbrockmendel committed:
REF: remove iloc case from _convert_slice_indexer

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into positional_slice

jbrockmendel committed:
REF: Move Loc-only methods to Loc (#31589)

* REF: Move Loc-only methods to Loc

* rebase fixup

* merge fixup

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into post-31786

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into post-31786

jbrockmendel committed:
Fix incorrect internal usage of loc, xref #31810

jbrockmendel committed:
BUG: loc.__setitem__ incorrectly accepting positional slices

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into post-31786

jorisvandenbossche commented:
Since this has been such a long time behavior (also used in our own tests ..), I think we should maybe consider raising a warning first.

jbrockmendel commented:
> I think we should maybe consider raising a warning first.

Yah, i'm leaning towards a deprecation cycle, too.

jbrockmendel committed:
update tests

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into post-31786

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into post-31786

jreback commented:
can you rebase, and agree with @jorisvandenbossche this is worth deprecating (even though its an incorrect behavior)

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into post-31786

jbrockmendel committed:
deprecate instead of raise

jbrockmendel committed:
whatsnew

jbrockmendel commented:
updated per comments, but i think the "Do X instead" part isnt quite right.  "Use iloc instead" works for Series, but doesnt work as well for DataFrame

jreback commented:
lgtm. can you add to the deprcations issue  & rebase; ping on green.

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into post-31786

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into post-31786

jbrockmendel committed:
update test

jbrockmendel commented:
added to the deprecation tracker, rebased, and green (though the CI is broken for unrelated reasons)

However I would still like to get a better message for users suggesting an alternative to use for the DataFrame case.  cc @jorisvandenbossche for suggestions?

jbrockmendel commented:
@WillAyd suggestions for a more helpful message for the 2D case?

TomAugspurger commented on pandas/core/indexes/base.py:
```diff
@@ -3177,8 +3177,18 @@ def is_int(v):
+            if kind == "loc":
+                # GH#16121, GH#24612, GH#31810
+                warnings.warn(
+                    "Slicing .loc with a positional slice is not supported, "
```
```suggestion
                    "Slicing a positional slice with .loc is not supported, "
```

TomAugspurger reviewed (commented):
The 2D case is something like this?

```python
df.iloc[:4, ['A', 'B'] = 1
```

I'd say something like "use .loc with labels, or .iloc with positions instead." I don't think we can really say which will be easier for the user in the error message, so just don't take a position.

jorisvandenbossche commented:
> I'd say something like "use .loc with labels, or .iloc with positions instead." I don't think we can really say which will be easier for the user in the error message, so just don't take a position.

+1 on that. Trying to explain both options more in detail will get a bit complex for the deprecation warning I think

jbrockmendel committed:
Update pandas/core/indexes/base.py

Co-Authored-By: Tom Augspurger <TomAugspurger@users.noreply.github.com>

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into post-31786

jbrockmendel committed:
Merge branch 'post-31786' of github.com:jbrockmendel/pandas into post-31786

jbrockmendel committed:
update message

jbrockmendel commented:
> I'd say something like "use .loc with labels, or .iloc with positions instead." I don't think we can really say which will be easier for the user in the error message, so just don't take a position.

Makes sense, updated.

jreback reviewed (approved):
lgtm. unless @jorisvandenbossche or @TomAugspurger have other comments, merge away

jorisvandenbossche commented on pandas/tests/indexing/test_loc.py:
```diff
@@ -975,3 +976,42 @@ def test_loc_mixed_int_float():
+    with pytest.raises(TypeError, match=msg):
+        df.loc[1:3, 1]
+
+    with tm.assert_produces_warning(FutureWarning, check_stacklevel=False):
```
I would optimise the stacklevel so it works correctly for dataframe, instead of series. 
I would think doing this with dataframe will be more common (it's also the case that is more difficult to solve). At least in our tests it was mainly with dataframes

jbrockmendel replied:
updated stacklevel

jorisvandenbossche reviewed (changes requested):
Only one comment about the stacklevel, for the rest looks good

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into post-31786

jbrockmendel committed:
optimize stacklevel for DataFrame

jbrockmendel committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into post-31786

jreback commented:
thanks @jbrockmendel
